open-apparel-registry/open-apparel-registry
# Add trigram match on facility name fallback when matching via API

## Overview

This new feature of the facility matching/creation API addresses problems experienced by an MSI when attempting to match records with partial or incomplete addresses by adding opt-in fallback text matching on facility name filtered by country when there are no dedupe matches.

Connects #1094 


## Notes

Installing the pg_trgm extension "[adds] functions and operators for determining the similarity of alphanumeric text based on trigram matching"

https://postgresql.org/docs/current/pgtrgm.html

We use the same 0.50 confidence threshold as our dedupe matcher for consistency. It is a reasonable starting point and could be adjusted in the future.

We are making this an opt-in feature by adding a new optional query string argument so all existing behavior is preserved.

## Testing Instructions

* Run `./scripts/resetdb` and verify that it completes without error
* Browse http://localhost:8081/api/docs/#!/facilities/facilities_create
* Authorize swagger with the development testing token `[Token 1d18b962d6f976b0b7e8fcf9fcc39b56cf278051]` 
* Set the `create` option to false and POST this JSON. Verify that a normal `MATCHED` is returned with a non-zero confidence score
```
[{
    "country": "VN",
    "name": "BI (VN) Co. Ltd.",
    "address": "Vietnam"
}]
```
* POST this example and verify that there is no match (NEW_FACILITY) is returned
```
{
    "country": "VN",
    "name": "BI Co. Ltd.",
    "address": "Vietnam"
}
```
* Set the `textonlyfallback` option to `true` and POST the same example. Verify that `POTENTIAL_MATCH` items are returned with a confidence score of 0 and `text_only_match` set to true.
```
{
    "country": "VN",
    "name": "BI Co. Ltd.",
    "address": "Vietnam"
}
```
* Set the `create` option to `true` and submit the same item. Verify that that the matches in the response contain `confirm_match_url` and `reject_match_url` links.
```
{
    "country": "VN",
    "name": "BI Co. Ltd.",
    "address": "Vietnam"
}
```
* Copy the ID from the first `confirm_match_url`, browse http://localhost:8081/api/docs/#!/facility-matches/facility_matches_confirm and submit the ID. Verify a successful response and copy the matched OAR ID
* Browse localhost:6543/facilities/{id} and verify that the new match is listed.

## Checklist
